### ISSUE TITLE
"iptables-logger" job not working on Jammy stemcell

### ISSUE THREAD

**Issue by jochenehret** [cloudfoundry/silk-release#67]
## Issue

The "iptables-logger" job doesn't work with the default settings on Jammy stemcells.

## Context

Tested silk-release 3.13.0 on `bosh-aws-xen-hvm-ubuntu-jammy-go_agent` version 1.18. This default setting is not valid on Jammy:
https://github.com/cloudfoundry/silk-release/blob/8139eff188a3a7f7b4f71a2c381c70f6f698dfa4/jobs/iptables-logger/spec#L20

I don't know however if this log file has just been disabled or moved to another folder.

## Steps to Reproduce

Deploy CF with Jammy stemcell, e.g. from this cf-deployment revision:
https://github.com/cloudfoundry/cf-deployment/commit/b901faebbbe53f4367ca04d99a5f5b5708a03d21
Add this ops file to enable the iptables-logger job:
https://github.com/cloudfoundry/cf-deployment/blob/develop/operations/experimental/enable-iptables-logger.yml

The update of the "diego-cell" VM should fail. On the cell, check the log `/var/vcap/sys/log/iptables-logger/iptables-logger.stderr.log`:
```
goroutine 1 [running]:
code.cloudfoundry.org/lager.(*logger).Fatal(0xc0000683c0, {0x8085fb, 0xa}, {0x8a3a80?, 0xc00007fc50}, {0x0, 0x0, 0xc000124228?})
        /var/vcap/data/compile/iptables-logger/src/code.cloudfoundry.org/vendor/code.cloudfoundry.org/lager/logger.go:162 +0x5dc
main.main()
        /var/vcap/data/compile/iptables-logger/src/code.cloudfoundry.org/iptables-logger/cmd/iptables-logger/main.go:74 +0x423
panic: open /var/log/kern.log: no such file or directory
```

## Expected result

The iptables-logger job should work.

## Current result

The iptables-logger job fails.

## Possible Fix

If the log is somewhere else, it would be sufficient to change the default value:
https://github.com/cloudfoundry/silk-release/blob/8139eff188a3a7f7b4f71a2c381c70f6f698dfa4/jobs/iptables-logger/spec#L20

I'm trying to find out if this is a problem with the Jammy stemcell.

## Additional Context

See above.

**Comment by ramonskie** [comment]
maby we can even log the iptables to a different file
i found https://dracocybersecurity.com/configure-rsyslog-to-log-iptables-log-to-another-log-file-instead-of-kern-log-syslog-log/
let us know if anyone else is interested in this

**Comment by ameowlia** [comment]
👋 Hi @jochenehret, Thank you for reporting this issue.

I have not seen this error in our pipeline...yet.

Our pipeline setup
* Stemcell: Jammy GCP Light 0.1.10
* Iptables logging enabled

I just bumped our pipeline to use the GCP light 1.18. I will keep you updated on the results.

**Comment by ameowlia** [comment]
Hi @jochenehret,

I was not able to reproduce this bug with the GCP light 1.18 stemcell. Maybe it is related only to the AWS stemcell?

I will continue to investigate.

**Comment by ameowlia** [comment]
Bosh team members @klakin-pivotal and @abg looked into this issue and found that rsyslog is running on GCP jammy vms but not on any AWS vms.

```
# gcp
# systemctl status rsyslog
Active: active (running)


# aws
# systemctl status rsyslog
Active: inactive (dead)
```

They concluded that this is likely a bug because `rsyslogd` isn't being started on Bosh or on the VMs it deploys on AWS with the 1.18 stemcell.

They are looking into a fix. I will provide a link here to that story when I get one.

Given that we think this is a stemcell bug and it is only affecting AWS, I suggest either:
1. disabling iptables logging and moving forward. (If you do this, make sure to add a known issue to the release note in cf-d).
2. waiting for a stemcell fix.

**Comment by klakin-pivotal** [comment]
@klakin-pivotal should be tagged for work stuff. That other guy is a personal account that doesn't do work stuff.

**Comment by Benjamintf1** [comment]
Another workaround, based on my understanding, may be to deploy syslog-release on your vms. It should run sysctl restart rsyslog, and as far as I know, cause rsyslog to start correctly. You can also use it to collect system logs(or run with "disabled" set to true). A opsfile can be found here https://github.com/cloudfoundry/cf-deployment/blob/main/operations/addons/enable-component-syslog.yml

**Comment by jochenehret** [comment]
Hi everyone,

currently we don't have the "enable-component-syslog" ops file configured for the experimental validation: https://github.com/cloudfoundry/cf-deployment/blob/6ec2aca405f23a2eb32ec4108d3385edcfdb9b22/ci/pipelines/cf-deployment.yml#L1075

So far the experimental validation succeeded for all stemcells and all IaaSs. So this looks like a regression on the AWS Jammy stemcell. Let's wait for @klakin-pivotal and @abg if this can be fixed in the AWS stemcell.

**Comment by ramonskie** [comment]
i have found some regressions and todo's that should have been fixed with the jammy stemcell.
all related to syslog. not sure if it fixes it
on of it is this commit https://github.com/cloudfoundry/bosh-linux-stemcell-builder/commit/4b05d5d1154c97207f51cfb09644364b472766f0
which will bump rsyslog to the latest version. (which we also use with bionic/xenial)

the other one is https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/ubuntu-jammy/master/stemcell_builder/stages/rsyslog_config/apply.sh#L49-L54

**Comment by jochenehret** [comment]
Ok so we have the following options:
1. Wait for a fix for the AWS Jammy stemcell (so that silk-release works again with its default options).
2. Use the "enable-component-syslog" ops file to enable syslog manually (this would then be a prerequisite for deploying the iptables-logger on AWS Jammy)
3. Simply remove the "enable-iptables-logger" from the "experimental" validation as that feature doesn't work anymore out-of-the box.

So if we can expect a fix for 1. in the next days, I would go for that option. Otherwise, I would tend to option 3.

**Comment by geofffranks** [comment]
@jochenehret it looks like this is fixed in the jammy v1.44 stemcell. Have you been able to confirm that?

**Comment by tpbrisco** [comment]
I'm not sure if this helps or not - but I was able to find (what appears to be) a race condition between the iptables-logger job, and the runsv job for rsyslog.

I observed a iptables-logger deployment fail on a single diego cell (two succeeded, one failed).  Looking at the failed diego-cell, it did not have a /var/log/kern.log file, and which was causing the iptables-logger monit job to fail/loop/lock (the monit/bpm job never reported failure).

- Rsyslog, when running, creates the /var/log/kern.log file (presumably copying the systemd journal into there) based on the /etc/rsyslog.d/50-default.conf.
- the iptables-logger wants to read the /var/log/kern.log -- if that file is not present, it appears to hang for some reason.
The BOSH deployment job was never completing, because the iptables-logger job was not failing - it appeared to be locked up / unable to fail gracefully.

Terminating the iptables-logger process prejudicially allowed the bpm job to fail, and the BOSH deployment to fail.  Once terminated, poking at the runsv rsyslog job indicated that it was now creating the /var/log/kern.log file, and a monit restart of iptables came up successfully.

The iptables-logger job appeared to be waiting on the rsyslogd "kern.log" file, but the rsyslog was not creating that file.  It did create that file once the iptables-logger was killed.

The deadlock could presumably be avoided with either (a) a simple wait loop in the monit start script for the /var/log/kern.log, or (b) a more graceful/backoff from the iptables-logger binary.

Environment: VirtualBox 6.1
Stemcell bosh-warden-boshlite-ubuntu-jammy-go_agent, 1.360
cf-deployment, 3 diego cells, with experimental iptables-logger opsfile

Recreate the issue:
- monit stop iptables-logger
- sv down rsyslog
- rm /var/log/kern.log
- monit start iptables-logger
- sv up rsyslog
Observe:
monit output that iptables-logger eventually moves to "not monitored" state, and /var/log/kern.log never appears.  A "sv status" indicates that rsyslog is running.
Resolve:
monit stop iptables-logger
touch /var/log/kern.log
chmod syslog:syslog /var/log/kern.log  (or wait, data may eventually appear)
monit start iptables-logger

It _could_ be related to the rsyslog job never creating the kern.log file - though I do not see what circumstances would lead to that.
